txpool/txpool.go

```diff
@@ -471,60 +510,195 @@ func (t *TxPool) ResetWithHeader(h *types.Header) {
+	nextNonce uint64, // The valid nonce (reference for pruning)
+	cleanupCallback func(txn *types.Transaction), // Additional cleanup logic
+) {
+	// Prune out all the now possibly low-nonce transactions in the promoted queue
+	t.pendingQueue.lock.Lock()
+
+	// Find the txns that correspond to this account
+	droppedPendingTxs := 0
+	for _, pendingQueueTxn := range t.pendingQueue.index {
+		// Check if the txn in the promoted queue matches the search criteria
+		if pendingQueueTxn.from == address && // The sender of this txn is the account we're looking for
+			pendingQueueTxn.tx.Nonce < nextNonce { // The nonce on this promoted txn is invalid
+			// Transaction found, drop it from the pending queue
+			if dropped := t.pendingQueue.dropTx(pendingQueueTxn.tx); dropped {
+				// Update the log data
+				droppedPendingTxs++
+				t.logger.Debug(
+					fmt.Sprintf(
+						"Dropping promoted txn [%s]",
+						pendingQueueTxn.tx.Hash.String(),
+					),
+				)
+
+				cleanupCallback(pendingQueueTxn.tx)
 			}
 		}
 	}
 
-	delTxns := map[types.Hash]*types.Transaction{}
+	t.pendingQueue.lock.Unlock()
+
+	// Print out the number of dropped pending txns
+	t.logger.Debug(
+		fmt.Sprintf(
+			"Dropped %d promoted txns for account [%s]",
+			droppedPendingTxs,
+			address.String(),
+		),
+	)
+}
+
+// extractTransactions Groups the transactions by account and queries the state
+// for the latest nonce data
+func (t *TxPool) extractTransactions(evnt *blockchain.Event) map[types.Address]*processEventWrapper {
+	// Instantiate the account process event wrapper map.
+	// It is used for grouping transactions on an account basis
+	// and for storing next state nonce information for easy lookup
+	// Account address -> processEventWrapper
+	eventWrapperMap := make(map[types.Address]*processEventWrapper)
+
+	// Grab the latest state root now that the block has been inserted
+	stateRoot := t.store.Header().StateRoot
+
+	// Keep track of all the transactions
 	for _, evnt := range evnt.NewChain {
-		// remove these transactions from the pool
+		// Grab the block that has just been written to state
 		block, ok := t.store.GetBlockByHash(evnt.Hash, true)
 		if !ok {
 			t.logger.Error("block not found on txn del", "hash", block.Hash())
 		} else {
+			// Compile transactions that should be accounted for in the TxPool
 			for _, txn := range block.Transactions {
-				delete(addTxns, txn.Hash)
-				delTxns[txn.Hash] = txn
+				// Save the transaction
+				eventWrapper, wrapperFound := eventWrapperMap[txn.From]
+				if !wrapperFound {
+					// Initialize the wrapper
+					eventWrapper = &processEventWrapper{
+						// Grab the latest nonce from state
+						stateNonce: t.store.GetNonce(stateRoot, txn.From),
+						// Set up the transaction array
+						transactions: make([]*types.Transaction, 0),
+					}
+
+					eventWrapperMap[txn.From] = eventWrapper
+				}
+
+				// Add the transaction to the wrapper
+				eventWrapper.addTxn(txn)
```

Comment:
Maybe add a `continue`  in the case of `if !ok { ... }` and lose the else branch, so to make it less verbose.

Reply:
Fixed 👍 

4d1fb51